File: ai_agent/backend/mqtt_client.py
"""
mqtt_client.py
--------------
Dedicated MQTT client module.

- Connects to the broker and subscribes to all sensor topics.
- Maintains a live in-memory snapshot of the latest sensor readings.
- Exposes publish_command() for the agent to send relay control messages.
- Never crashes the server on broker unavailability — sets a flag instead.
"""

import json
import logging
import threading
import paho.mqtt.client as mqtt

from config import MQTT_BROKER, MQTT_PORT

logger = logging.getLogger(__name__)

# ── In-memory state ──────────────────────────────────────────────────────────
# This dict is updated every time a new MQTT message arrives (~every 2 seconds
# from the ESP8266). The REST API reads from here to serve the dashboard.
latest_sensor_data: dict = {
    "temp": None,
    "humidity": None,
    "ldr": None,
    "pir": None,
    "relay": None,
    "mqtt_connected": False,
}

_client: mqtt.Client | None = None
_lock = threading.Lock()  # Thread-safe writes to latest_sensor_data


# ── Callbacks ────────────────────────────────────────────────────────────────

def _on_connect(client, userdata, flags, rc):
    if rc == 0:
        logger.info("✅ MQTT connected to %s:%s", MQTT_BROKER, MQTT_PORT)
        with _lock:
            latest_sensor_data["mqtt_connected"] = True
        # Subscribe to all sensor topics
        client.subscribe([
            ("home/ldr",   2),
            ("home/pir",   2),
            ("home/dht",   2),
            ("home/relay", 2),
        ])
    else:
        logger.warning("⚠️  MQTT connection failed with code %s", rc)
        with _lock:
            latest_sensor_data["mqtt_connected"] = False


def _on_disconnect(client, userdata, rc):
    logger.warning("⚠️  MQTT disconnected (rc=%s). Will auto-reconnect.", rc)
    with _lock:
        latest_sensor_data["mqtt_connected"] = False


def _on_message(client, userdata, msg):
    topic = msg.topic
    payload = msg.payload.decode("utf-8").strip()

    with _lock:
        try:
            if topic == "home/ldr":
                latest_sensor_data["ldr"] = round(float(payload), 2)

            elif topic == "home/pir":
                latest_sensor_data["pir"] = int(payload)

            elif topic == "home/dht":
                data = json.loads(payload)
                latest_sensor_data["temp"] = round(float(data.get("temp", 0)), 2)
                latest_sensor_data["humidity"] = round(float(data.get("hum", 0)), 2)

            elif topic == "home/relay":
                latest_sensor_data["relay"] = payload  # "ON" or "OFF"

        except (ValueError, json.JSONDecodeError, KeyError) as e:
            logger.error("❌ Failed to parse message on %s: %s | Error: %s", topic, payload, e)


# ── Public API ────────────────────────────────────────────────────────────────

def start():
    """Initialize and connect the MQTT client in a background thread."""
    global _client
    try:
        _client = mqtt.Client(client_id="IoT_AI_Agent_Server")
        _client.on_connect    = _on_connect
        _client.on_disconnect = _on_disconnect
        _client.on_message    = _on_message

        _client.connect_async(MQTT_BROKER, MQTT_PORT, keepalive=60)
        _client.loop_start()  # Non-blocking background thread
        logger.info("🔌 MQTT client started. Connecting to %s...", MQTT_BROKER)

    except Exception as e:
        # Do NOT raise — server should start even if MQTT is temporarily down
        logger.error("❌ MQTT client failed to start: %s", e)
        with _lock:
            latest_sensor_data["mqtt_connected"] = False


def stop():
    """Cleanly disconnect the MQTT client."""
    global _client
    if _client:
        _client.loop_stop()
        _client.disconnect()
        logger.info("🔌 MQTT client disconnected.")


def publish_command(topic: str, message: str) -> dict:
    """
    Publish a control command to the MQTT broker.
    Returns a result dict so the agent can report success/failure naturally.
    """
    if _client is None or not latest_sensor_data.get("mqtt_connected"):
        return {"success": False, "error": "MQTT broker is not connected. Command not sent."}

    try:
        result = _client.publish(topic, message, qos=1)
        if result.rc == mqtt.MQTT_ERR_SUCCESS:
            logger.info("📤 Published '%s' to '%s'", message, topic)
            return {"success": True, "topic": topic, "message": message}
        else:
            return {"success": False, "error": f"Publish failed with code {result.rc}"}
    except Exception as e:
        logger.error("❌ Publish error: %s", e)
        return {"success": False, "error": str(e)}

File: ai_agent/backend/mqttClient.js
import mqtt from 'mqtt';
import { MQTT_BROKER, MQTT_PORT } from './config.js';

export const latestSensorData = {
  temp: null,
  humidity: null,
  ldr: null,
  pir: null,
  relay: null,
  mqtt_connected: false,
};

let client = null;

export function start() {
  try {
    const brokerUrl = `mqtt://${MQTT_BROKER}:${MQTT_PORT}`;
    console.log(`🔌 MQTT client starting. Connecting to ${brokerUrl}...`);

    client = mqtt.connect(brokerUrl, {
      clientId: 'IoT_AI_Agent_Server_Node',
      clean: true,
      connectTimeout: 5000,
      reconnectPeriod: 5000,
    });

    client.on('connect', () => {
      console.log(`✅ MQTT connected to ${MQTT_BROKER}:${MQTT_PORT}`);
      latestSensorData.mqtt_connected = true;

      // Subscribe to all sensor topics
      client.subscribe([
        'home/ldr',
        'home/pir',
        'home/dht',
        'home/relay'
      ], { qos: 2 }, (err) => {
        if (err) {
          console.error('❌ MQTT subscription error:', err);
        }
      });
    });

    client.on('offline', () => {
      console.warn('⚠️  MQTT client offline. Attempting reconnect...');
      latestSensorData.mqtt_connected = false;
    });

    client.on('close', () => {
      latestSensorData.mqtt_connected = false;
    });

    client.on('error', (err) => {
      console.error('❌ MQTT client error:', err.message);
      latestSensorData.mqtt_connected = false;
    });

    client.on('message', (topic, message) => {
      const payload = message.toString().trim();

      try {
        if (topic === 'home/ldr') {
          latestSensorData.ldr = Math.round(parseFloat(payload) * 100) / 100;
        } else if (topic === 'home/pir') {
          latestSensorData.pir = parseInt(payload, 10);
        } else if (topic === 'home/dht') {
          const data = JSON.parse(payload);
          latestSensorData.temp = Math.round(parseFloat(data.temp || 0) * 100) / 100;
          latestSensorData.humidity = Math.round(parseFloat(data.hum || 0) * 100) / 100;
        } else if (topic === 'home/relay') {
          latestSensorData.relay = payload; // "ON" or "OFF"
        }
      } catch (err) {
        console.error(`❌ Failed to parse message on ${topic}: ${payload} | Error: ${err.message}`);
      }
    });

  } catch (error) {
    console.error(`❌ MQTT client failed to start: ${error.message}`);
    latestSensorData.mqtt_connected = false;
  }
}

export function stop() {
  if (client) {
    client.end(false, () => {
      console.log('🔌 MQTT client cleanly disconnected.');
    });
    latestSensorData.mqtt_connected = false;
  }
}

export function publishCommand(topic, message) {
  if (!client || !latestSensorData.mqtt_connected) {
    return { success: false, error: 'MQTT broker is not connected. Command not sent.' };
  }

  return new Promise((resolve) => {
    client.publish(topic, message, { qos: 1 }, (err) => {
      if (err) {
        console.error(`❌ Publish error: ${err.message}`);
        resolve({ success: false, error: err.message });
      } else {
        console.log(`📤 Published '${message}' to '${topic}'`);
        resolve({ success: true, topic, message });
      }
    });
  });
}
